### ISSUE TITLE
Robot operating system (ROS) interface

### ISSUE THREAD

**Issue by peabody124** [TauLabs/TauLabs#1408]
I'm curious about the utility of a ROS interface to our system. I can envision two
approaches. The first (and easiest) being using a light weight GCS (e.g. #1407)
to provide a ROS node that supports a series of messages relevant for quadcopters.

The later (and more powerful?) would be to actually use rosserial to implement
a module that handles low level ROS messages. This would allow directly treating
the quadcopter as a standalone node.

I'm still starting to read through the ROS documentation and it isn't clear to me what
the general set of messages you would want to have in order to achieve something
useful.

**Comment by ernieift** [comment]
I've got an Edison Kit last week. It can be a nice processor update for this task. It could powered by a single USART port and also communicate over it or simply use an USB port. For the first steps we could use our telemetry protocol. I'll dig a little deeper...

**Comment by peabody124** [comment]
Awesome. I have one too as well as freedom which can run similar software. Similar thoughts were crossing my mind.

**Comment by TSC21** [comment]
Please check this: http://forum.taulabs.org/viewtopic.php?f=18&t=575&p=4674#p4674

**Comment by peabody124** [comment]
@kubark42 BTW is this something you guys have ever discussed?

**Comment by kubark42** [comment]
No, not really. We evaluated ROS, but our feeling was that it still needed some time to mature before it's ready for widespread use outside the lab.

**Comment by peabody124** [comment]
@kubark42 yeah I'm not sure how much t would benefit end users. However, just seems like it would open a lot of doors for playing and researchers. I keep feeling the urge to do this and then keep wondering the best way.

**Comment by TSC21** [comment]
Well in my case I moved to MAVLink-based autopilots and I'm working with [mavros](http://wiki.ros.org/mavros), which works as an MAVLink extension ROS node. On the forum, I also made reference to [opgateway](https://github.com/vooon/opgateway), from the same creator of `mavros`, which started but didn't finished the project as working with UAVTalk isn't is pleasant as working with MAVLink as is right now.

**Comment by TSC21** [comment]
In any case, I tried to pick on `opgateway` and continue the dev but quit after sometime for lack of support. So if anyone is interested, maybe we can give it a shot and proceed with the development.

**Comment by TSC21** [comment]
@kubark42 maybe we can work together on a solution?

**Comment by mlyle** [comment]
@tsc21 -- would love to work together on ROS stuff.  I'm working over on dRonin, but anything you'd like to do would probably work well on both Tau & dRonin now.  We've enhanced the python uavtalk API some which would provide a good foundation to build from.

**Comment by TSC21** [comment]
What do you have in mind?

**Comment by peabody124** [comment]
@TSC21 so the big question is always have in mind is whether to implement a ROS node directly on the flight controller, versus write some ground software that acts as a proxy. Since there is always a computer for ROS the later seems ok but inelegant and will add non trivial latency. It looks like writing a node would require linking in some C++ code, but should be doable and with F4 processor should easily fit.

**Comment by peabody124** [comment]
Note that this would require using rosserial http://wiki.ros.org/rosserial so there is still that requirement on the ground side.

**Comment by mlyle** [comment]
Not really sure there's a latency difference between serializing uavtalk messages vs. serializing other messages.  I think the approach I'd prefer is to implement a thin layer on top of the dRonin python telemetry modules that runs as a ROS node.

**Comment by TSC21** [comment]
> implement a ROS node directly on the flight controller

I don't think that is actually possible unless the flight controller runs a Linux-based OS instead of a RTOS, which I think is not the case.
So the best approach I think is write a wrapper on the onboard computer/ground station side to convert UAVO to ROS msgs. That wrapper is basically a node that is listening/writing to a serial connection to the FCU. `mavros` itself is a pretty neat implementation as a Mavlink<->ROS bridge, so something similar may become the best thing to do. And for that, I think the best way to go is proceed with the development of `opgateway`.

**Comment by peabody124** [comment]
@TSC21 no it is definitely possible. that's how people can get even things like arduino to talk directly to ROS. you still need rosserial to bridge to TCP. the big up side of this is
- compatibility since you won't need an additional layer to talk to the FC
- on the FC side you can optimize the messages so that it using the bandwidth more effectively and lower latency (versus having to deal with the whole UAVTalk telemetry and having that stream)
- makes it easier to write custom code on the FC too

the down side is that it is more work, may (or may not) require linking in C++ (I'm looking at few different microcontroller implementations), and will require a bit more on the FC so probably will only work on F4 boards (although you probably want something with built in telemetry so that's not a huge deal).

I'm working on the former for now, although we'll see if it dead ends.

**Comment by TSC21** [comment]
> and will require a bit more on the FC so probably will only work on F4 boards

For me that is a big issue, at least from a user/researcher perspective right? I think the idea is not overcharge the FCU with more computational expensive stuff but to retrieve anything possible from the FCU and run it on a companion board. A well implemented communication stack would have a lower latency.
So these are two approaches, and I'm feeling that I will want to tackle the one which I know works for sure and with acceptable/good results.

**Comment by peabody124** [comment]
Well it's probably more of a flash space issue. For example Sparky1 is pretty tight on that basis. However, the F4 has more than enough spare cycles to do it and won't be limiting.

From my experience using UAVTalk for control here https://vimeo.com/73060129 the latency was enough to make controlling definitely harder. I'd prefer to avoid the middleman if possible and suspect it will work better.

**Comment by peabody124** [comment]
I got C++ linking to Sparky2 and with mostly stub code it doesn't increase the binary size appreciably (312776 to 312944)

**Comment by peabody124** [comment]
![screen shot 2016-02-21 at 5 01 48 pm](https://cloud.githubusercontent.com/assets/618346/13206091/25769d80-d8bd-11e5-8e62-0bf90131bc91.png)

Got it working (just basic text node). FW size is 326728 so not a huge bump

**Comment by peabody124** [comment]
And here is visualization of pose

https://www.instagram.com/p/BCFwLMGlquZ

so working pretty well so far
